File: 03/solution.rb
test = %w(
  00100
  11110
  10110
  10111
  10101
  01111
  00111
  11100
  10000
  11001
  00010
  01010
)

input_file = File.open('./input.txt', 'r')
input = input_file.readlines(chomp: true)
input_file.close

def count(lines)
  line_length = lines[0].length
  counts = *(0..line_length-1).to_a.fill(0)
  lines.each do |line|
    line.split('').each_with_index do |char, i|
      if char == "1"
        counts[i] += 1
      end
    end
  end

  counts
end

winning_count = input.length / 2
counts = count(input)

@most_common = counts.map{|c| if c>=winning_count then 1 else 0 end }.join('')
@least_common = counts.map{|c| if c<=winning_count then 1 else 0 end }.join('')
gamma = @most_common.to_i(2)
epsilon = @least_common.to_i(2) 

puts "part one: ", gamma * epsilon
puts ''

winning_count = test.length / 2
counts = count(test)
def most_common(lines, pos)
  count = lines.map{ |l| l.split('')[pos].to_i }.sum
  if count >= lines.length.to_f/2
    1
  else
    0
  end
end
def least_common(lines, pos)
  most_common(lines, pos)^0b1
end

def ox_rating(lines, i = 0)
  if lines.length == 1
    return lines[0].to_i(2)
  end

  filter = most_common(lines, i)
  filtered_lines = lines.select{ |line| line.split('')[i].to_i == filter } 
  return ox_rating(filtered_lines, i+1)
end

def co2_rating(lines, i = 0)
  if lines.length == 1
    return lines[0].to_i(2)
  end

  filter = least_common(lines, i)
  filtered_lines = lines.select{ |line| line.split('')[i].to_i == filter } 
  return co2_rating(filtered_lines, i+1)
end

puts "part two:", ox_rating(input)*co2_rating(input)

File: 03/solution.exs
defmodule Input do
  def lines(filename) do
    File.read(filename)
    |> then(fn {:ok, contents} -> contents end)
    |> String.split("\n", trim: true)
  end
end

defmodule DayThree do
  def matrix(lines) do
    lines
    |> Enum.map(&String.split(&1, "", trim: true))
    |> Enum.map(&Enum.map(&1, fn s -> String.to_integer(s) end))
  end

  def count(lines, i) do
    lines
    |> matrix
    |> Enum.map(&Enum.at(&1, i))
    |> Enum.sum
  end

  def most_common(lines, i) do
    lines
    |> count(i)
    |> then(fn sum -> 
      case sum >= length(lines) / 2 do
        true -> 1
        false -> 0
      end
    end)
  end

  def least_common(lines, i) do
    Bitwise.bxor(most_common(lines, i), 1)
  end

  def gamma(lines) do
    0..String.length(List.first(lines))-1
    |> Enum.to_list
    |> Enum.map(fn i -> most_common(lines, i) end)
    |> Enum.join
    |> String.to_integer(2)
  end

  def epsilon(lines) do
    0..String.length(List.first(lines))-1
    |> Enum.to_list
    |> Enum.map(fn i -> least_common(lines, i) end)
    |> Enum.join
    |> String.to_integer(2)
  end

  def one(lines) do
    gamma(lines) * epsilon(lines)
  end

  def o2_rating(lines, i) when length(lines) > 1 do
    filter = most_common(lines, i)
    filtered_lines = lines
      |> Enum.filter(fn line ->
        line
        |> String.split("", trim: true)
        |> Enum.at(i)
        |> String.to_integer
        |> then(fn x -> x == filter end)
      end)
    o2_rating(filtered_lines, i+1)
  end
  def o2_rating(lines, i) when length(lines) == 1 do
    lines
    |> Enum.at(0)
    |> String.to_integer(2) 
  end

  def co2_rating(lines, i) when length(lines) > 1 do
    filter = least_common(lines, i)
    filtered_lines = lines
      |> Enum.filter(fn line ->
        line
        |> String.split("", trim: true)
        |> Enum.at(i)
        |> String.to_integer
        |> then(fn x -> x == filter end)
      end)
    co2_rating(filtered_lines, i+1)
  end
  def co2_rating(lines, i) when length(lines) == 1 do
    lines
    |> Enum.at(0)
    |> String.to_integer(2) 
  end

  def two(lines) do
    co2_rating(lines, 0) * o2_rating(lines, 0)
  end
end

Input.lines('input.txt')
|> DayThree.one
|> IO.inspect

Input.lines('input.txt')
|> DayThree.two
|> IO.inspect
